cairo/ethereum/cancun/trie.cairo

```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
-//         // if prefix_length == 0:
-//         // break
-//             // break
-//     // if prefix_length > 0:
-//     // prefix = arbitrary_key[level:level + prefix_length]
-//     // return ExtensionNode(prefix, encode_internal_node(patricialize(obj, level + prefix_length)))
-//         // prefix = arbitrary_key[level:level + prefix_length]
-//         // return ExtensionNode(prefix, encode_internal_node(patricialize(obj, level + prefix_length)))
-//     // branches: List[MutableMapping[Bytes, Bytes]] = []
-//     // for _ in range(16):
-//     // branches.append({})
-//         // branches.append({})
-//     // value = b''
-//     // for key in obj:
-//     // if len(key) == level:
-//     // if isinstance(obj[key], (Account, Receipt, Uint)):
-//     // raise AssertionError
-//     // value = obj[key]
-//     // else:
-//     // branches[key[level]][key] = obj[key]
-//         // if len(key) == level:
-//         // if isinstance(obj[key], (Account, Receipt, Uint)):
-//         // raise AssertionError
-//         // value = obj[key]
-//         // else:
-//         // branches[key[level]][key] = obj[key]
-//             // if isinstance(obj[key], (Account, Receipt, Uint)):
-//             // raise AssertionError
-//                 // raise AssertionError
-//             // value = obj[key]
-//         // else:
-//             // branches[key[level]][key] = obj[key]
-//     // return BranchNode([encode_internal_node(patricialize(branches[k], level + 1)) for k in range(16)], value)
-// }
+func _search_common_prefix_length{
+    range_check_ptr, substring: Bytes, level: Uint, dict_ptr_stop: BytesBytesDictAccess*
+}(obj: BytesBytesDictAccess*, current_length: felt) -> felt {
+    alloc_locals;
+    if (obj == dict_ptr_stop) {
+        return current_length;
+    }
+
+    tempvar sliced_key = Bytes(
+        new BytesStruct(obj.key.value.data + level.value, obj.key.value.len - level.value)
+    );
+    let result = common_prefix_length(substring, sliced_key);
+    let current_length = min(result, current_length);
+    if (current_length == 0) {
+        return 0;
+    }
+
+    return _search_common_prefix_length(obj + BytesBytesDictAccess.SIZE, current_length);
+}
+
+// @dev Iterate over the DictAccesses from dict_ptr_start to dict_ptr to create a new DictAccess*
+// @dev whose key[level] = nibble.
+func _get_branche_for_nibble_at_level(obj: MappingBytesBytes, nibble: felt, level: felt) -> (
+    MappingBytesBytes, Bytes
+) {
+    alloc_locals;
+    let (local branch: BytesBytesDictAccess*) = alloc();
+    local dict_ptr_stop: BytesBytesDictAccess* = obj.value.dict_ptr;
+    local value: Bytes;
+    local value_set: felt;
+
+    tempvar branch = branch;
+    tempvar dict_ptr = obj.value.dict_ptr_start;
+
+    loop:
+    let branch = cast([ap - 2], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 1], BytesBytesDictAccess*);
+    // The verifier just needs to make sure that whatever case we are in is properly asserted.
```

Comment:
```suggestion
    // The assertions on the values returned by this nondet memory assignment are verified in the corresponding branches
```

Reply:
I think that it's important to strength the fact that the verifier doesn't know nor need to verify at each iteration which branch is the right one, but only that whatever branch the prover chose is sufficiently constraint